genetics/src/main/java/binnie/genetics/recipes/RotatedOreIngredient.java

```diff
@@ -51,7 +51,7 @@ public RotatedOreIngredient(String ore, int offset) {
 					lst.add(itemstack);
 			}
 			Collections.rotate(lst, offset);
-			this.array = lst.toArray(new ItemStack[lst.size()]);
+			this.array = lst.toArray(new ItemStack[0]);
```

Comment:
I don't understand why this line was changed.

Reply:
> One might think that it's preferable to pass a pre-sized array instead of a throwaway zero-length array. [Aleksey Shipilev wrote an article analyzing this question.](https://shipilev.net/blog/2016/arrays-wisdom-ancients/#_conclusion) The answer is, somewhat counterintuitively, that creating a zero-length array is potentially faster.

> Briefly, the reason is that allocation is cheap, a zero-length array is small, and it's probably going to be thrown away and garbage collected quickly, which is also cheap. By contrast, creating a pre-sized array requires it to be allocated and then filled with nulls/zeroes. It's then passed to toArray(), which then fills it with values from the list. Thus, every array element is typically written twice. By passing a zero-length array to toArray(), this allows the array allocation to occur in the same code as the array filling code, providing the opportunity for the JIT compiler to bypass the initial zero-fill, since it knows that every array element will be filled.

https://stackoverflow.com/questions/37625877/could-java-util-arraylistt-toarray-be-made-friendlier

Reply:
Thanks, that makes sense.